File: to-dos/ui/src/utils/priority.ts
/**
 * to-dos 优先排位分算法 (v0.5.0)
 *
 * 前端版本 - TypeScript 实现
 * 与后端 priority.py 算法保持一致
 */

import type { Todo, Category } from '@/types'

/**
 * 解析 ISO 时间字符串
 */
function parseTime(timeStr: string | undefined): Date | null {
  if (!timeStr) return null
  try {
    const d = new Date(timeStr)
    return isNaN(d.getTime()) ? null : d
  } catch {
    return null
  }
}

/**
 * 计算优先排位分
 */
export function calcPriorityScore(
  todo: Todo,
  category: Category,
  now: Date = new Date(),
): number {
  const deadlineDt = parseTime(todo.deadline)
  const startDt = parseTime(todo.start_time || todo.created_at)

  // 校验：已过期
  if (deadlineDt && deadlineDt <= now) return -1

  // 校验：无效时间范围
  if (!deadlineDt || !startDt || deadlineDt <= startDt) return 0

  const difficulty = category.difficulty ?? 5
  const priority = todo.priority_rank ?? 0

  // 时间差（分钟）
  const totalTime = (deadlineDt.getTime() - startDt.getTime()) / 60000
  const remainingTime = (deadlineDt.getTime() - now.getTime()) / 60000

  // 评估时间（对数域）
  let logEvalTime: number
  if (difficulty === 0) {
    logEvalTime = 0
  } else {
    const base = 1 - Math.pow(50 / (difficulty + 50), priority)
    logEvalTime = base <= 0
      ? Math.log(0.001)
      : Math.log(base) + Math.log(totalTime)
  }

  // 评估充裕度
  const logEvalSufficiency = Math.log(Math.max(remainingTime, 0.001)) - logEvalTime

  // 标定时间
  const estimatedTime = todo.estimated_time || todo.time_estimate || 60
  const calibrateTime = Math.min(totalTime, estimatedTime)
  const logCalibrateSufficiency =
    Math.log(Math.max(remainingTime, 0.001)) - Math.log(Math.max(calibrateTime, 0.001))

  // 时间充裕度
  const evalSuff = Math.exp(logEvalSufficiency)
  const calibrateSuff = Math.exp(logCalibrateSufficiency)
  const timeSufficiency =
    0.65 * Math.exp(0.5 * logEvalSufficiency + 0.5 * logCalibrateSufficiency) +
    0.7 * evalSuff * calibrateSuff / (evalSuff + calibrateSuff + 0.001)

  // 基础优先度
  const importanceScore = todo.important ? (145 + difficulty) : 0
  const urgencyScore = todo.urgent ? (1335 + 3 * difficulty) : 1000
  const basePriority = (importanceScore + urgencyScore) / Math.max(timeSufficiency, 0.001)

  // 优先排位分
  const effectivePriority = Math.max(priority, 1)
  let multiplier = Math.pow(1.06, effectivePriority) + 1.35 * Math.pow(effectivePriority, 1.5) - 1
  if (multiplier <= 0) multiplier = 0.001

  const logScore = Math.log(Math.max(basePriority, 0.001)) + Math.log(Math.max(multiplier, 0.001))
  let score = Math.exp(logScore)

  // 软封顶
  if (score > 1e10) {
    score = 1e10 + 100 * Math.log10(score)
  }

  return Math.round(score)
}

/**
 * 格式化分数显示
 */
export function formatScoreDisplay(score: number): string {
  if (score < 0) return '过期'
  if (score === 0) return '0'
  if (score >= 1e10) {
    const number = (score - 1e10) / 100
    return `EL${number.toFixed(1)}`
  }
  if (score >= 1e8) {
    const number = Math.log10(score)
    return `EL${number.toFixed(1)}`
  }
  return String(score)
}

/**
 * 计算分类分数
 * sum(该分类下所有待办的分数) / sqrt(待办数量)
 */
export function calcCategoryScores(
  todos: Todo[],
  categories: Category[],
  now: Date = new Date(),
): Map<string, number> {
  const categoryMap = new Map(categories.map(c => [c.id, c]))
  const scores = new Map<string, number>()
  const counts = new Map<string, number>()

  for (const todo of todos) {
    if (['completed', 'cancelled', 'archived'].includes(todo.status)) continue

    const category = categoryMap.get(todo.category)
    if (!category) continue

    const score = calcPriorityScore(todo, category, now)
    scores.set(todo.category, (scores.get(todo.category) ?? 0) + score)
    counts.set(todo.category, (counts.get(todo.category) ?? 0) + 1)
  }

  const result = new Map<string, number>()
  for (const [catId, totalScore] of scores) {
    const count = counts.get(catId) ?? 1
    result.set(catId, totalScore / Math.sqrt(count))
  }

  return result
}

/**
 * 批量计算所有待办的分数
 */
export function calcAllScores(
  todos: Todo[],
  categories: Category[],
  now: Date = new Date(),
): Map<string, number> {
  const categoryMap = new Map(categories.map(c => [c.id, c]))
  const scores = new Map<string, number>()

  for (const todo of todos) {
    const category = categoryMap.get(todo.category)
    if (category) {
      scores.set(todo.id, calcPriorityScore(todo, category, now))
    }
  }

  return scores
}

File: to-dos/src/priority.py
"""
to-dos 优先排位分算法模块 (v0.5.0)

核心算法：计算待办事项的优先排位分，用于动态排序

特点：
- 全程使用对数域计算，避免大数溢出
- 综合考虑优先级、紧急度、重要度、时间充裕度
- 支持实时计算和格式化显示
"""

import math
from datetime import datetime
from typing import Optional

from .types import Todo, Category
from .utils import TimeHelper


def calc_priority_score(
    todo: Todo,
    category: Category,
    now: Optional[datetime] = None,
) -> int:
    """
    计算优先排位分，使用对数域避免大数

    参数：
        todo: 待办事项
        category: 所属分类
        now: 当前时间，默认为 datetime.now()

    返回：
        优先排位分（整数）

    算法说明：
        1. 校验：已过期返回 -inf，无效时间范围返回 0
        2. 计算时间差（分钟）：total_time, remaining_time
        3. 评估时间（对数域）：基于难度和优先级
        4. 评估充裕度：剩余时间 vs 评估时间
        5. 标定充裕度：剩余时间 vs 预估时间
        6. 时间充裕度：几何平均 + 调和平均组合
        7. 基础优先度：基于重要度和紧急度
        8. 优先排位分：对数域计算最终分数
        9. 软封顶：超过 1e10 使用对数压缩
    """
    if now is None:
        now = datetime.now()

    # 解析时间
    deadline_dt = TimeHelper.parse_iso(todo.deadline) if todo.deadline else None
    start_time_str = todo.start_time or todo.created_at
    start_dt = TimeHelper.parse_iso(start_time_str)

    # 校验：已过期
    if deadline_dt and deadline_dt <= now:
        return -1  # 已过期，返回 -1 表示最低优先级

    # 校验：无效时间范围
    if not deadline_dt or not start_dt or deadline_dt <= start_dt:
        return 0

    # 难度和优先级
    difficulty = category.difficulty
    priority = todo.priority_rank

    # 时间差（分钟）
    total_time = (deadline_dt - start_dt).total_seconds() / 60
    remaining_time = (deadline_dt - now).total_seconds() / 60

    # 评估时间（对数域）
    if difficulty == 0:
        log_eval_time = 0  # 瞬间完成
    else:
        base = 1 - (50 / (difficulty + 50)) ** priority
        if base <= 0:
            log_eval_time = math.log(0.001)  # 避免 log(0)
        else:
            log_eval_time = math.log(base) + math.log(total_time)

    # 评估充裕度
    log_eval_sufficiency = math.log(max(remaining_time, 0.001)) - log_eval_time

    # 标定时间
    estimated_time = todo.estimated_time or todo.time_estimate or 60
    calibrate_time = min(total_time, estimated_time)
    log_calibrate_sufficiency = (
        math.log(max(remaining_time, 0.001)) - math.log(max(calibrate_time, 0.001))
    )

    # 时间充裕度（几何平均 + 调和平均组合）
    eval_suff = math.exp(log_eval_sufficiency)
    calibrate_suff = math.exp(log_calibrate_sufficiency)

    time_sufficiency = (
        0.65 * math.exp(0.5 * log_eval_sufficiency + 0.5 * log_calibrate_sufficiency)
        + 0.7 * eval_suff * calibrate_suff / (eval_suff + calibrate_suff + 0.001)
    )

    # 基础优先度
    importance_score = (145 + difficulty) if todo.important else 0
    urgency_score = (1335 + 3 * difficulty) if todo.urgent else 1000
    base_priority = (importance_score + urgency_score) / max(time_sufficiency, 0.001)

    # 优先排位分（对数域）
    multiplier = 1.06 ** priority + 1.35 * (priority ** 1.5) - 1
    if multiplier <= 0:
        multiplier = 0.001

    log_score = math.log(max(base_priority, 0.001)) + math.log(max(multiplier, 0.001))
    score = math.exp(log_score)

    # 软封顶（对数压缩）
    if score > 1e10:
        score = int(1e10 + 100 * math.log10(score) + 0.5)
    else:
        score = int(score + 0.5)

    return score


def format_score_display(score: int) -> str:
    """
    格式化分数显示

    参数：
        score: 优先排位分

    返回：
        格式化后的分数字符串

    规则：
        - score >= 1e10: 显示为 EL{number}，number = (score - 1e10) / 100
        - score >= 1e8: 显示为 EL{log10(score)}
        - score < 1e8: 直接显示数字
        - score < 0: 显示 "过期"
    """
    if score < 0:
        return "过期"
    elif score == 0:
        return "0"
    elif score >= 1e10:
        number = (score - 1e10) / 100
        return f"EL{number:.1f}"
    elif score >= 1e8:
        number = math.log10(score)
        return f"EL{number:.1f}"
    else:
        return str(score)


def calc_category_score(
    todos: list,
    categories: list,
    now: Optional[datetime] = None,
) -> dict:
    """
    计算每个分类的优先排位分

    参数：
        todos: 待办列表
        categories: 分类列表
        now: 当前时间

    返回：
        分类 ID 到分数的映射 {category_id: score}

    算法：
        分类分数 = sum(该分类下所有待办的分数) / sqrt(待办数量)
    """
    if now is None:
        now = datetime.now()

    # 构建分类映射
    category_map = {c.id: c for c in categories}

    # 计算每个分类的分数
    category_scores = {}
    category_counts = {}

    for todo in todos:
        # 跳过已完成、已取消、已归档的待办
        if todo.status.value in ('completed', 'cancelled', 'archived'):
            continue

        cat_id = todo.category
        category = category_map.get(cat_id)
        if not category:
            continue

        score = calc_priority_score(todo, category, now)

        if cat_id not in category_scores:
            category_scores[cat_id] = 0
            category_counts[cat_id] = 0

        category_scores[cat_id] += score
        category_counts[cat_id] += 1

    # 计算最终分数：sum / sqrt(count)
    result = {}
    for cat_id, total_score in category_scores.items():
        count = category_counts[cat_id]
        if count > 0:
            result[cat_id] = total_score / math.sqrt(count)
        else:
            result[cat_id] = 0

    return result


def calc_all_scores(
    todos: list,
    categories: list,
    now: Optional[datetime] = None,
) -> dict:
    """
    计算所有待办的优先排位分

    参数：
        todos: 待办列表
        categories: 分类列表
        now: 当前时间

    返回：
        待办 ID 到分数的映射 {todo_id: score}
    """
    if now is None:
        now = datetime.now()

    # 构建分类映射
    category_map = {c.id: c for c in categories}

    scores = {}
    for todo in todos:
        category = category_map.get(todo.category)
        if category:
            scores[todo.id] = calc_priority_score(todo, category, now)
        else:
            scores[todo.id] = 0

    return scores
